includes/subscriptions/class-wc-payments-subscriptions-event-handler.php

```diff
@@ -0,0 +1,194 @@
+		if ( 0 === $subscription->get_time( 'next_payment' ) ) {
+			// TODO: Add error handling to these {cancel/suspend}_subscription calls i.e. add a subscription order note if the WCPay subscription wasn't cancelled.
+			if ( ! $subscription->has_status( 'on-hold' ) && 0 !== $subscription->get_time( 'end' ) ) {
+				$this->subscription_service->cancel_subscription( $subscription );
+			} else {
+				$this->subscription_service->suspend_subscription( $subscription );
+			}
+		} else {
+			// Translators: %s Scheduled/upcoming payment date in Y-m-d H:i:s format.
+			$subscription->add_order_note( sprintf( __( 'Next automatic payment scheduled for %s.', 'woocommerce-payments' ), get_date_from_gmt( gmdate( 'Y-m-d H:i:s', $wcpay_subscription['current_period_end'] ), wc_date_format() . ' ' . wc_time_format() ) ) );
+
+			// Update the subscription in WC to match the WCPay Subscription's next payment date.
+			$this->subscription_service->update_dates_to_match_wcpay_subscription( $wcpay_subscription, $subscription );
+
+			$response = $this->invoice_service->create_invoice_items_for_subscription( $subscription, $wcpay_customer_id, $wcpay_subscription_id );
+
+			if ( is_wp_error( $response ) ) {
+				throw new Rest_Request_Exception( $response->get_error_message() );
+			}
+		}
+	}
+
+	/**
+	 * Renews a subscription associated with paid invoice.
+	 *
+	 * @param array $body The event body that triggered the webhook.
+	 *
+	 * @throws Rest_Request_Exception Required parameters not found.
+	 */
+	public function handle_invoice_paid( array $body ) {
+		$event_data            = $this->get_event_property( $body, 'data' );
+		$event_object          = $this->get_event_property( $event_data, 'object' );
+		$wcpay_subscription_id = $this->get_event_property( $event_object, 'subscription' );
+		$wcpay_invoice_id      = $this->get_event_property( $event_object, 'id' );
+		$subscription          = WC_Payments_Subscription_Service::get_subscription_from_wcpay_subscription_id( $wcpay_subscription_id );
+
+		if ( ! $subscription ) {
+			throw new Rest_Request_Exception( __( 'Cannot find subscription for the incoming "invoice.paid" event.', 'woocommerce-payments' ) );
+		}
+
+		// This incoming invoice.paid event is linked to the subscription parent invoice and can be ignored.
+		if ( WC_Payments_Invoice_Service::get_subscription_invoice_id( $subscription ) === $wcpay_invoice_id ) {
+			return;
+		}
+
+		$order = wc_get_order( WC_Payments_Invoice_Service::get_order_id_by_invoice_id( $wcpay_invoice_id ) );
+
+		if ( ! $order ) {
+			$order = wcs_create_renewal_order( $subscription );
+
+			if ( is_wp_error( $order ) ) {
+				throw new Rest_Request_Exception( __( 'Unable to generate renewal order for subscription on the "invoice.paid" event.', 'woocommerce-payments' ) );
+			} else {
+				$this->invoice_service->set_order_invoice_id( $order, $wcpay_invoice_id );
+				$order->payment_complete();
+			}
+		}
+
```

Comment:
This is just an observation, so no change needs to be made for this.

I did some thinking about how we handle the end of subscriptions. Right now, we cancel the subscription at `invoice.upcoming` when WC doesn't expect the next payment date. But what happens if the store goes down and can't respond to that webhook prior to the actual renewal date in Stripe?

I'm wondering if it would be better to [set_pending_cancel_for_subscription](https://github.com/Automattic/woocommerce-payments/blob/be7a0b69cba684a7de58155944c8bb2487eb3ec2/includes/subscriptions/class-wc-payments-subscription-service.php#L229) at `invoice.paid` at the penultimate renewal?

This is probably something we can discuss and address outside of this PR, but wanted to note this anyway since I just thought about it.

Reply:
This is a good suggestion!
I've just created an issue for this: https://github.com/Automattic/woocommerce-payments/issues/2860

Feel free to add any notes to that issue in case I've misunderstood this comment 😅